weave/trace/concurrent/futures.py

```diff
@@ -0,0 +1,244 @@
+"""ThreadPoolExecutor for executing tasks in parallel.
+
+This module provides a FutureExecutor class for asynchronous task execution in Weave.
+It offers a simple interface for deferring tasks and chaining operations on futures,
+which is particularly useful for I/O-bound operations or long-running tasks that
+shouldn't block the main execution thread.
+
+The FutureExecutor class includes two main methods:
+1. defer: For submitting tasks to be executed asynchronously.
+2. then: For chaining operations on futures, allowing for easy composition
+   of asynchronous workflows.
+
+These utilities are designed to work seamlessly with Weave's tracing and
+context management systems, ensuring that asynchronous operations maintain
+the necessary context and can be properly traced and monitored.
+
+Usage of this module can significantly improve the performance and
+responsiveness of Weave applications by allowing concurrent execution
+of independent tasks, especially in scenarios involving network requests,
+file I/O, or other operations with significant waiting times.
+
+To use this module, create an instance of FutureExecutor and use its methods
+to manage asynchronous tasks:
+
+    executor = FutureExecutor()
+    future = executor.defer(some_function, arg1, arg2)
+    result_future = executor.then([future], process_result)
+"""
+
+import atexit
+import concurrent.futures
+import logging
+from concurrent.futures import Future, wait
+from contextvars import ContextVar
+from threading import Lock
+from typing import Any, Callable, List, Optional, TypeVar
+
+from weave.trace.context import get_raise_on_captured_errors
+from weave.trace.util import ContextAwareThreadPoolExecutor
+
+logger = logging.getLogger(__name__)
+
+# Constants
+THREAD_NAME_PREFIX = "WeaveThreadPool"
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+class FutureExecutor:
+    """A utility for thread-local threadpool execution and promise-like chaining.
+
+    This class provides a thread-safe way to submit and execute jobs asynchronously
+    using thread-local ThreadPoolExecutors. It ensures proper shutdown of the executors when the
+    object is deleted or when the program exits.
```

Comment:
Why did you choose threadpool instead of using asyncio?

Reply:
A few important reasons:
1. Thread jobs start immediately after submission, rather than waiting for an `await` call
2. You can use threads anywhere (including within asyncio coroutines), whereas asyncio coroutines can only be "awaited" within an `async` context
3. Personally, threads are easier to reason about in terms of execution flow.

Threads are the OGs, asyncio have additional constraints (namely #2) that make it challenging. I would hate to have to rework all our code (and our customer's code) just to get a defered value.

Reply:
Ok, that's fair.  Do you expect end-users to ever have to use this executor, or will it be mainly internal (including but not exceeding the integrations surface)?

Reply:
Also, have you seen AnyIO?
https://anyio.readthedocs.io/en/stable/

This is what I instinctually jumped to when I saw this problem.